File: sdk/python/src/spendguard/pricing.py
"""Pricing lookup helper — Phase 4 O4 USD-denominated budget support.

Adapters that want to charge an LLM call against a USD-denominated
budget need a way to convert (provider, model, token_counts) into
USD-micros. This module exposes a small `PricingLookup` that takes a
frozen pricing table (matching the seed YAML in
`deploy/demo/init/pricing/seed.yaml`) and computes µUSD per call.

For production, the pricing table comes from
`spendguard_canonical.pricing_table` keyed by the bundle's
`pricing_version`. POC sidecar adapters hardcode a subset; a future
slice will surface the full table via the sidecar handshake or a
control-plane API.

Usage::

    from spendguard.pricing import PricingLookup, USD_MICROS_PER_USD

    pricing = PricingLookup({
        ("openai", "gpt-4o-mini", "input"):  0.15,    # $/1M tokens
        ("openai", "gpt-4o-mini", "output"): 0.60,
        ("anthropic", "claude-haiku-4-5-20251001", "input"):  1.00,
        ("anthropic", "claude-haiku-4-5-20251001", "output"): 5.00,
    })

    micros = pricing.usd_micros_for_call(
        provider="openai", model="gpt-4o-mini",
        input_tokens=120, output_tokens=40,
    )
    # → math.ceil((120 * 0.15 + 40 * 0.60) / 1e6 * 1e6) → 42 µUSD = $0.000042
"""

from __future__ import annotations

import math
from collections.abc import Mapping

USD_MICROS_PER_USD = 1_000_000

PriceKey = tuple[str, str, str]  # (provider, model, token_kind)
PriceTable = Mapping[PriceKey, float]


class PricingLookup:
    """Frozen-at-construction pricing table → USD-micros computation.

    The lookup is *not* side-effecting — there is no DB query, no
    network call. Callers fetch / cache pricing once (e.g., at
    handshake time) and pass it in.
    """

    def __init__(self, table: PriceTable, *, default_kind: str = "output") -> None:
        self._table = dict(table)
        self._default_kind = default_kind

    def price_per_million(
        self, provider: str, model: str, token_kind: str
    ) -> float | None:
        """Return $/1M-tokens or None if (provider, model, kind) is missing."""
        return self._table.get((provider, model, token_kind))

    def usd_micros_for_call(
        self,
        *,
        provider: str,
        model: str,
        input_tokens: int = 0,
        output_tokens: int = 0,
        cached_input_tokens: int = 0,
    ) -> int:
        """Compute the µUSD cost of a single LLM call.

        Charges per-kind when prices are available; falls back to the
        default kind (typically `output`) for any token bucket without
        a configured price. Round up to the nearest µUSD so the
        customer is never under-charged due to floating-point
        truncation.
        """
        usd = 0.0
        for kind, count in (
            ("input", input_tokens),
            ("output", output_tokens),
            ("cached_input", cached_input_tokens),
        ):
            if count <= 0:
                continue
            price = self.price_per_million(provider, model, kind)
            if price is None:
                price = self.price_per_million(
                    provider, model, self._default_kind
                ) or 0.0
            usd += count * price / 1_000_000
        # Round up to the nearest µUSD (fail-safe for the customer).
        return max(1, math.ceil(usd * USD_MICROS_PER_USD))


__all__ = ["PriceKey", "PriceTable", "PricingLookup", "USD_MICROS_PER_USD"]

File: sdk/typescript/src/pricing.ts
// SpendGuard SDK — pricing lookup + USD-micros computation (SLICE 6 /
// COV_S05_06).
//
// Mirror of `sdk/python/src/spendguard/pricing.py`. Adapters that want to
// charge an LLM call against a USD-denominated budget convert
// (provider, model, token_counts) into USD-micros here.
//
// For production, the pricing table comes from the canonical pricing-table
// keyed by the bundle's `pricing_version` (forward-reserved; sidecar fetch
// not in v0.1.x). POC sidecar adapters hardcode a subset; the demo snapshot
// ships under `@spendguard/sdk/pricing/demo` as the embedded `DEMO_PRICING`
// `PricingLookup` instance.
//
// Spec refs:
//   - design.md §4.9 (LOCKED `PricingLookup` surface)
//   - implementation.md §8 (`src/pricing.ts`)
//   - tests.md §3.4 (pricing computation matrix)

import { PricingMissingError } from "./errors.js";

/** USD micros per USD — 1 USD = 1,000,000 µUSD. */
export const USD_MICROS_PER_USD = 1_000_000;

/**
 * Composite lookup key for the pricing table.
 *
 * `(provider, model, tokenKind)` triples key the `PriceTable`. `tokenKind`
 * is one of `"input"`, `"output"`, `"cached_input"`, `"vision_input"`,
 * `"audio_input"`, `"reasoning"`.
 *
 * The `PriceTable` Map serializes the triple to a `${provider}|${model}|${kind}`
 * string — JS Map keys do not support tuple identity, so we flatten.
 */
export type PriceKey = readonly [provider: string, model: string, tokenKind: string];

/**
 * Frozen-at-construction pricing table.
 *
 * Keys: `${provider}|${model}|${tokenKind}` strings.
 * Values: USD price per million tokens (e.g. `0.15` = $0.15 / 1M tokens).
 */
export type PriceTable = ReadonlyMap<string, number>;

/**
 * Frozen pricing table → USD-micros computation.
 *
 * The lookup is not side-effecting — there is no DB query, no network call.
 * Callers fetch + cache pricing once (e.g., at handshake time) and pass it in.
 * Mirrors Python `PricingLookup` semantics including:
 *   - default kind fallback (typically `"output"`) when a token kind has no
 *     configured price.
 *   - per-kind charging for input / output / cached_input buckets.
 *   - round-up to the nearest µUSD so the customer is never under-charged.
 */
export class PricingLookup {
  private readonly table: PriceTable;
  private readonly defaultKind: string;

  constructor(table: PriceTable, opts?: { defaultKind?: string }) {
    this.table = table;
    this.defaultKind = opts?.defaultKind ?? "output";
  }

  /** Return $/1M-tokens or `null` if `(provider, model, kind)` is missing. */
  pricePerMillion(provider: string, model: string, tokenKind: string): number | null {
    const v = this.table.get(`${provider}|${model}|${tokenKind}`);
    return v === undefined ? null : v;
  }

  /**
   * Compute the µUSD cost of a single LLM call.
   *
   * Charges per-kind when prices are available; falls back to the default
   * kind (typically `"output"`) for any token bucket without a configured
   * price. Result is rounded UP to the nearest µUSD (fail-safe for the
   * customer — never under-charge due to FP truncation). The minimum
   * returned value is `1` (we never claim zero cost on a non-zero token
   * count).
   *
   * Fail-closed: when a token bucket has a non-zero count but NEITHER the
   * specific kind NOR the default kind has a configured price, this throws
   * {@link PricingMissingError} instead of silently charging $0. Coercing a
   * missing price to 0 would under-count the budget — the exact under-charge
   * the guardrail exists to prevent — and unknown/new models are precisely
   * the ones most likely to be mispriced. Buckets with a zero count never
   * trigger this (no charge is attributed, so a missing price is irrelevant).
   *
   * @throws {@link PricingMissingError} when a non-zero token bucket has no
   *   resolvable price.
   */
  usdMicrosForCall(args: {
    provider: string;
    model: string;
    inputTokens?: number;
    outputTokens?: number;
    cachedInputTokens?: number;
  }): number {
    let usd = 0;
    const charge = (kind: string, count: number): void => {
      if (count <= 0) return;
      // Preserve the existing default-kind fallback; only the FINAL `?? 0`
      // coercion is removed — a still-missing price now fails loud.
      const p =
        this.pricePerMillion(args.provider, args.model, kind) ??
        this.pricePerMillion(args.provider, args.model, this.defaultKind);
      if (p === null) {
        throw new PricingMissingError({
          provider: args.provider,
          model: args.model,
          tokenKind: kind,
        });
      }
      usd += (count * p) / 1_000_000;
    };
    charge("input", args.inputTokens ?? 0);
    charge("output", args.outputTokens ?? 0);
    charge("cached_input", args.cachedInputTokens ?? 0);
    if (usd <= 0) return 0;
    return Math.max(1, Math.ceil(usd * USD_MICROS_PER_USD));
  }
}
